File: src/main/claudeDesktop.js
'use strict';

/**
 * One-click integration with the Claude Desktop app's "Third-Party Inference".
 *
 * Unlike Claude Code (a plain ~/.claude/settings.json we write directly), Claude Desktop
 * (bundle `com.anthropic.claudefordesktop`) reads its third-party-inference settings from macOS
 * *Managed Preferences*, delivered as a Configuration Profile (.mobileconfig). So:
 *
 *   connect():  generate a profile pre-filled with the local ccbud gateway, then hand it to macOS.
 *               `profiles` CLI "no longer supports installs", so the user approves it once in
 *               System Settings › Profiles (admin password). This is NOT a risky operation — it
 *               only changes where inference is sent; macOS just requires you to confirm it.
 *   disconnect(): `profiles remove -identifier …` still works, run via an admin prompt → ~one-click
 *               restore. Falls back to opening System Settings if that path is unavailable.
 *
 * Schema (extracted from Claude Desktop's own profile generator):
 *   PayloadType (inner) = com.anthropic.claudefordesktop, with managed keys:
 *   inferenceProvider="gateway", inferenceCredentialKind="static",
 *   inferenceGatewayBaseUrl, inferenceGatewayApiKey, inferenceGatewayAuthScheme="bearer".
 */

const fs = require('fs');
const path = require('path');
const os = require('os');
const crypto = require('crypto');
const { exec, execFile, execFileSync } = require('child_process');
const { CLAUDE_TIER_MODELS } = require('./claudeModels');

const BUNDLE_ID = 'com.anthropic.claudefordesktop';
const PROFILE_IDENTIFIER = 'dev.ccbud.gateway.claude-desktop-inference';
const PROFILES_PANE = 'x-apple.systempreferences:com.apple.preferences.configurationprofiles';

const isMac = () => process.platform === 'darwin';
const endpoint = (port) => `http://localhost:${port || 8788}`;
const profilePath = () =>
  path.join(os.homedir(), '.ccbud', 'claude-desktop-inference.mobileconfig');

function appInstalled() {
  if (!isMac()) return false;
  const candidates = [
    '/Applications/Claude.app',
    path.join(os.homedir(), 'Applications', 'Claude.app'),
    path.join(os.homedir(), 'Library', 'Application Support', 'Claude'),
  ];
  return candidates.some((p) => { try { return fs.existsSync(p); } catch (_) { return false; } });
}

// Stable UUIDs so re-generating the profile updates it in place rather than duplicating.
function uuidFrom(seed) {
  const h = crypto.createHash('sha1').update(String(seed)).digest('hex');
  return `${h.slice(0, 8)}-${h.slice(8, 12)}-${h.slice(12, 16)}-${h.slice(16, 20)}-${h.slice(20, 32)}`.toUpperCase();
}
const xmlEsc = (s) => String(s).replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;');

function buildProfile(port, token) {
  // Claude Desktop's Gateway picker needs an explicit model list, stored as a SINGLE JSON string.
  // Names carry Anthropic keywords so its client-side validation accepts them, and they match what
  // ccbud returns from /v1/models; the gateway then tier-maps each onto the active provider.
  const inferenceModels = JSON.stringify(CLAUDE_TIER_MODELS.map((m) => Object.assign(
    { name: m.name, anthropicFamilyTier: m.tier }, m.familyDefault ? { isFamilyDefault: true } : {},
  )));
  const settings = {
    inferenceProvider: 'gateway',
    inferenceCredentialKind: 'static',
    inferenceGatewayBaseUrl: endpoint(port),
    inferenceGatewayApiKey: token || 'ccbud-local',
    inferenceGatewayAuthScheme: 'bearer',
    inferenceModels,
  };
  const body = Object.entries(settings)
    .map(([k, v]) => `      <key>${k}</key>\n      <string>${xmlEsc(v)}</string>`).join('\n');
  return `<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
  <key>PayloadContent</key>
  <array>
    <dict>
      <key>PayloadType</key>
      <string>${BUNDLE_ID}</string>
      <key>PayloadIdentifier</key>
      <string>${PROFILE_IDENTIFIER}.settings</string>
      <key>PayloadUUID</key>
      <string>${uuidFrom(PROFILE_IDENTIFIER + '.settings')}</string>
      <key>PayloadVersion</key>
      <integer>1</integer>
      <key>PayloadDisplayName</key>
      <string>Claude Desktop Third-Party Inference (CC Buddy)</string>
${body}
    </dict>
  </array>
  <key>PayloadDisplayName</key>
  <string>CC Buddy · Claude Desktop 第三方推理</string>
  <key>PayloadDescription</key>
  <string>将 Claude 桌面版的模型推理指向本地 CC Buddy 网关（${endpoint(port)}）。可随时移除以还原为官方推理。</string>
  <key>PayloadIdentifier</key>
  <string>${PROFILE_IDENTIFIER}</string>
  <key>PayloadOrganization</key>
  <string>CC Buddy</string>
  <key>PayloadRemovalDisallowed</key>
  <false/>
  <key>PayloadScope</key>
  <string>User</string>
  <key>PayloadType</key>
  <string>Configuration</string>
  <key>PayloadUUID</key>
  <string>${uuidFrom(PROFILE_IDENTIFIER)}</string>
  <key>PayloadVersion</key>
  <integer>1</integer>
</dict>
</plist>
`;
}

// Read the effective managed baseUrl Claude Desktop would use (mirrors how it reads managed prefs).
function managedBaseUrl() {
  let user = '';
  try { user = os.userInfo().username; } catch (_) {}
  const paths = [`/Library/Managed Preferences/${BUNDLE_ID}.plist`];
  if (user) paths.push(`/Library/Managed Preferences/${user}/${BUNDLE_ID}.plist`);
  for (const p of paths) {
    try {
      if (!fs.existsSync(p)) continue;
      const out = execFileSync('/usr/bin/plutil', ['-extract', 'inferenceGatewayBaseUrl', 'raw', '-o', '-', p], {
        encoding: 'utf8', timeout: 4000,
      }).trim();
      if (out) return out;
    } catch (_) {}
  }
  return null;
}

function status(port) {
  return {
    supported: isMac(),
    installed: appInstalled(),
    connected: isMac() && managedBaseUrl() === endpoint(port),
    endpoint: endpoint(port),
  };
}

// Write the profile and open it + the Profiles pane for the user to approve (one-time macOS step).
function connect(port, token) {
  if (!isMac()) return { ok: false, reason: 'unsupported' };
  if (!appInstalled()) return { ok: false, reason: 'notInstalled' };
  const file = profilePath();
  try {
    fs.mkdirSync(path.dirname(file), { recursive: true });
    fs.writeFileSync(file, buildProfile(port, token), 'utf8');
  } catch (e) {
    return { ok: false, reason: 'write', message: e.message };
  }
  exec(`/usr/bin/open ${JSON.stringify(file)}`, () => {
    setTimeout(() => exec(`/usr/bin/open ${JSON.stringify(PROFILES_PANE)}`, () => {}), 1200);
  });
  return { ok: true, needsApproval: true, path: file };
}

// Remove the profile via an admin prompt (~one-click); fall back to System Settings on failure.
function disconnect() {
  if (!isMac()) return Promise.resolve({ ok: false, reason: 'unsupported' });
  return new Promise((resolve) => {
    const osa = `do shell script "/usr/bin/profiles remove -identifier ${PROFILE_IDENTIFIER}" with administrator privileges`;
    execFile('/usr/bin/osascript', ['-e', osa], (err, _stdout, stderr) => {
      if (!err) { resolve({ ok: true, removed: true }); return; }
      const msg = String((stderr || '') + (err && err.message ? err.message : ''));
      if (/-128|User canceled/i.test(msg)) { resolve({ ok: false, cancelled: true }); return; }
      // CLI removal unavailable → open System Settings so the user can remove it manually.
      exec(`/usr/bin/open ${JSON.stringify(PROFILES_PANE)}`, () => {});
      resolve({ ok: true, removed: false, needsApproval: true });
    });
  });
}

module.exports = {
  appInstalled, status, connect, disconnect, buildProfile, profilePath,
  BUNDLE_ID, PROFILE_IDENTIFIER,
};

File: src-tauri/src/claudedesktop.rs
// Claude Desktop "Third-Party Inference" integration — Rust port of claudeDesktop.js.
//
// Claude Desktop reads inference settings from macOS Managed Preferences, delivered as a
// Configuration Profile (.mobileconfig). connect() generates the profile pre-filled with the
// local gateway and opens it for the user to approve; disconnect() removes it via an admin
// prompt. The profile advertises the claude-* tier models (matching /v1/models) so a fresh
// Claude Desktop can pick a model and drive the gateway with zero per-user setup.

#![allow(dead_code)]

use crate::gateway::CLAUDE_TIER_MODELS;
use serde_json::{json, Value};
use std::path::PathBuf;

const BUNDLE_ID: &str = "com.anthropic.claudefordesktop";
const PROFILE_IDENTIFIER: &str = "dev.ccbud.gateway.claude-desktop-inference";
const PROFILES_PANE: &str =
    "x-apple.systempreferences:com.apple.preferences.configurationprofiles";

fn home() -> PathBuf {
    std::env::var("HOME").map(PathBuf::from).unwrap_or_else(|_| PathBuf::from("."))
}
fn is_mac() -> bool {
    cfg!(target_os = "macos")
}
fn endpoint(port: u16) -> String {
    format!("http://localhost:{}", port)
}
pub fn profile_path() -> PathBuf {
    home().join(".ccbud").join("claude-desktop-inference.mobileconfig")
}

fn uuid_from(seed: &str) -> String {
    use sha1::{Digest, Sha1};
    let mut h = Sha1::new();
    h.update(seed.as_bytes());
    let hex = format!("{:x}", h.finalize());
    format!("{}-{}-{}-{}-{}", &hex[0..8], &hex[8..12], &hex[12..16], &hex[16..20], &hex[20..32]).to_uppercase()
}
fn xml_esc(s: &str) -> String {
    s.replace('&', "&amp;").replace('<', "&lt;").replace('>', "&gt;")
}

fn app_installed() -> bool {
    if !is_mac() {
        return false;
    }
    [
        PathBuf::from("/Applications/Claude.app"),
        home().join("Applications").join("Claude.app"),
        home().join("Library").join("Application Support").join("Claude"),
    ]
    .iter()
    .any(|p| p.exists())
}

pub fn build_profile(port: u16, token: &str) -> String {
    // Gateway picker needs an explicit model list as a SINGLE JSON string; names carry Anthropic
    // keywords (so its validation accepts them) and match what /v1/models returns.
    let models: Vec<Value> = CLAUDE_TIER_MODELS
        .iter()
        .map(|(name, tier)| {
            let mut m = serde_json::Map::new();
            m.insert("name".into(), json!(name));
            m.insert("anthropicFamilyTier".into(), json!(tier));
            if *tier == "sonnet" {
                m.insert("isFamilyDefault".into(), json!(true));
            }
            Value::Object(m)
        })
        .collect();
    let inference_models = serde_json::to_string(&models).unwrap_or_default();

    let settings = [
        ("inferenceProvider", "gateway".to_string()),
        ("inferenceCredentialKind", "static".to_string()),
        ("inferenceGatewayBaseUrl", endpoint(port)),
        ("inferenceGatewayApiKey", if token.is_empty() { "ccbud-local".to_string() } else { token.to_string() }),
        ("inferenceGatewayAuthScheme", "bearer".to_string()),
        ("inferenceModels", inference_models),
    ];
    let body = settings
        .iter()
        .map(|(k, v)| format!("      <key>{}</key>\n      <string>{}</string>", k, xml_esc(v)))
        .collect::<Vec<_>>()
        .join("\n");

    format!(
        r#"<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
  <key>PayloadContent</key>
  <array>
    <dict>
      <key>PayloadType</key>
      <string>{bundle}</string>
      <key>PayloadIdentifier</key>
      <string>{ident}.settings</string>
      <key>PayloadUUID</key>
      <string>{uuid_settings}</string>
      <key>PayloadVersion</key>
      <integer>1</integer>
      <key>PayloadDisplayName</key>
      <string>Claude Desktop Third-Party Inference (CC Buddy)</string>
{body}
    </dict>
  </array>
  <key>PayloadDisplayName</key>
  <string>CC Buddy · Claude Desktop 第三方推理</string>
  <key>PayloadDescription</key>
  <string>将 Claude 桌面版的模型推理指向本地 CC Buddy 网关（{ep}）。可随时移除以还原为官方推理。</string>
  <key>PayloadIdentifier</key>
  <string>{ident}</string>
  <key>PayloadOrganization</key>
  <string>CC Buddy</string>
  <key>PayloadRemovalDisallowed</key>
  <false/>
  <key>PayloadScope</key>
  <string>User</string>
  <key>PayloadType</key>
  <string>Configuration</string>
  <key>PayloadUUID</key>
  <string>{uuid_root}</string>
  <key>PayloadVersion</key>
  <integer>1</integer>
</dict>
</plist>
"#,
        bundle = BUNDLE_ID,
        ident = PROFILE_IDENTIFIER,
        uuid_settings = uuid_from(&format!("{}.settings", PROFILE_IDENTIFIER)),
        uuid_root = uuid_from(PROFILE_IDENTIFIER),
        body = body,
        ep = endpoint(port),
    )
}

fn managed_base_url() -> Option<String> {
    if !is_mac() {
        return None;
    }
    let user = std::env::var("USER").unwrap_or_default();
    let mut paths = vec![format!("/Library/Managed Preferences/{}.plist", BUNDLE_ID)];
    if !user.is_empty() {
        paths.push(format!("/Library/Managed Preferences/{}/{}.plist", user, BUNDLE_ID));
    }
    for p in paths {
        if !std::path::Path::new(&p).exists() {
            continue;
        }
        if let Ok(out) = std::process::Command::new("/usr/bin/plutil")
            .args(["-extract", "inferenceGatewayBaseUrl", "raw", "-o", "-", &p])
            .output()
        {
            let s = String::from_utf8_lossy(&out.stdout).trim().to_string();
            if !s.is_empty() {
                return Some(s);
            }
        }
    }
    None
}

pub fn status(port: u16) -> Value {
    json!({
        "supported": is_mac(),
        "installed": app_installed(),
        "connected": is_mac() && managed_base_url().as_deref() == Some(endpoint(port).as_str()),
        "endpoint": endpoint(port),
    })
}

pub fn connect(port: u16, token: &str) -> Value {
    if !is_mac() {
        return json!({ "ok": false, "reason": "unsupported" });
    }
    if !app_installed() {
        return json!({ "ok": false, "reason": "notInstalled" });
    }
    let file = profile_path();
    if let Some(dir) = file.parent() {
        let _ = std::fs::create_dir_all(dir);
    }
    if std::fs::write(&file, build_profile(port, token)).is_err() {
        return json!({ "ok": false, "reason": "write" });
    }
    let f = file.to_string_lossy().to_string();
    let _ = std::process::Command::new("/usr/bin/open").arg(&f).spawn();
    // Take the user to System Settings › Profiles shortly after, so they can approve it (matches
    // claudeDesktop.js — without this many users get stuck not knowing where to approve).
    std::thread::spawn(|| {
        std::thread::sleep(std::time::Duration::from_millis(1200));
        let _ = std::process::Command::new("/usr/bin/open").arg(PROFILES_PANE).spawn();
    });
    json!({ "ok": true, "needsApproval": true, "path": f })
}

pub fn disconnect() -> Value {
    if !is_mac() {
        return json!({ "ok": false, "reason": "unsupported" });
    }
    let osa = format!(
        "do shell script \"/usr/bin/profiles remove -identifier {}\" with administrator privileges",
        PROFILE_IDENTIFIER
    );
    match std::process::Command::new("/usr/bin/osascript").args(["-e", &osa]).output() {
        Ok(o) if o.status.success() => json!({ "ok": true, "removed": true }),
        Ok(o) => {
            // User canceled the admin-password prompt (-128 / "User canceled") → report it as
            // cancelled instead of pretending it still needs approval. Otherwise (CLI unavailable)
            // fall back to opening System Settings so the user can remove it manually.
            let stderr = String::from_utf8_lossy(&o.stderr);
            if stderr.contains("-128") || stderr.to_lowercase().contains("user canceled") {
                json!({ "ok": false, "cancelled": true })
            } else {
                let _ = std::process::Command::new("/usr/bin/open").arg(PROFILES_PANE).spawn();
                json!({ "ok": true, "removed": false, "needsApproval": true })
            }
        }
        Err(_) => {
            let _ = std::process::Command::new("/usr/bin/open").arg(PROFILES_PANE).spawn();
            json!({ "ok": true, "removed": false, "needsApproval": true })
        }
    }
}
